### PR TITLE
[NeoMathEngine] blob convolution backward algorithms (fixed)

### PR DESCRIPTION
**Pull request by favorart**
* Algo1 -- Decrease RAM consumption and avoid excess matrix transposing -- speed-up
* Algo2 -- Avoid excess matrix transposing -- speed-up
* Unit-tests fixed

### CODE REVIEW

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 1]
```diff
@@ -553,319 +535,348 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
-		if( freeTermData != nullptr ) {
-			// Set the free term
-			setVectorToMatrixRows( outputDataPtr + outputRow * output.Width() * vectorSize, output.Width(),
-				vectorSize, freeTermDataRaw );
-		} else {
-			vectorFill0( outputDataPtr + outputRow * output.Width() * vectorSize, output.Width() * vectorSize );
-		}
+	const int ringBufRowSize = source.Width() * filterObjectSize;
+	// Minimum possible size of the ring-buffer
+	const int ringBufNeedNumRows = std::max( 1, ( filter.Height() * desc.DilationHeight + 1 + desc.StrideHeight - 1 ) / desc.StrideHeight );
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: _[reminder]_ use `totalFilterHeight`

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 2]
```diff
@@ -553,319 +535,348 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
+				updateRingBuf( ringBuf, batch, sourceRowStart, prevBatchedSourceRowStart );
+
+				const int filterRowStart = std::max( 0, row + filterRowStartShift );
+				int sourceRow = 0;
+				for( int filterRow = filterRowEnd; filterRow >= filterRowStart; filterRow -= desc.StrideHeight, ++sourceRow ) {
+					const int tempOffset = ringBuf.GetSourceRowOffset( sourceRow ) * ringBufRowSize + filterRow * filterRowSize;
+					// The temp blob stores the filter rows multiplied by source; add them to the result rows in correct positions
+					const float* tempRowData = ringBuf.GetRaw( tempOffset );
+					for( int filterColumn = filterColMin; filterColumn <= filterColMax; filterColumn += desc.StrideWidth, tempRowData += filterObjectSize ) {
+						const int minFilterColumn0 = std::min( 0, filterColumn ), maxFilterColumn0 = std::max( 0, filterColumn );
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: _codestyle_ Every variable declaration must be on a separate line

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 1 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 2 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

**Review by FedyuninV — APPROVED** [review]

**Pull request merged**